lib/streamlit/fileManager.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: utf-8 -*-
+# Copyright 2018-2019 Streamlit Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from streamlit.logger import get_logger
+
+LOGGER = get_logger(__name__)
+
+class File(object):
+    """Thread-safe queue that smartly accumulates the report's messages."""
+
+    def __init__(self, widget_id, name, size, last_modified, chunks):
+
+        self.widget_id = widget_id
+        self.name = name
+        self.size = size
+        self.last_modified = last_modified
+        self.total_chunks = chunks
+        self.buffers = {}
+        self.data = None
+
+class FileManager(object):
+    def __init__(self):
+        self._file_list = {}
+
+    def delete_all_files(self):
+        for widget_id in list(self._file_list):
+            self.delete_file(widget_id)
+
+    def delete_file(self, widget_id):
+        del self._file_list[widget_id]
+
+    def locate_new_file(self, widget_id, name, size, last_modified, chunks):
+        file = self._file_list.get(widget_id)
+        if file != None:
+            self.delete_file(widget_id)
```

Comment:
The 3 lines above would be clearer if you replaced with this:

```py
if widget_id in self._file_list:
  self.delete_file(widget_id)
```

...because this actually reads like what you want to happen: "if the widget ID exists in the file list, delete the file with that widget ID".